pylzy/lzy/env/explorer/classify.py

```diff
@@ -199,6 +212,105 @@ def _classify_modules_without_distributions(
+
+            new_meta_packages = self.requirements_to_meta_packages.get(name, ())
+            for meta_package in new_meta_packages:
+                if meta_package.name not in seen_meta_packages:
+                    meta_packages.add(meta_package)
+
+        # step 1: finding all meta packages that requires any package
+        # we already found though module exploring
+        for package in packages:
+            add_meta_requirements(package.name)
+
+        # step 2: classify all found meta packages
+        while meta_packages:
+            meta_package = meta_packages.pop()
+            seen_meta_packages.add(meta_package.name)
+
+            package = self._classify_distribution(meta_package, set())
+
+            # step 3: if metapackage is found on pypi, just add it to result
+            # as usual PypiPackage
+            if isinstance(package, PypiDistribution):
+                result.add(package)
+                continue
+
+            # package var must include PypiDistribution | LocalDistribution
+            # so this will never assert
+            if not isinstance(package, LocalDistribution):
+                assert_never()
+
+            # step 4: if metapackage is not found on pypi,
+            # we must treat it as local distribution,
+            # but it lacking files by defenition.
+            # also, as it lacking files, we didn't found
+            # it's requirement packages through module exporing
+            requirements: List[str] = meta_package.requires or []
+            for requirement_string in requirements:
+                requirement_name = get_name_from_requirement_string(requirement_string)
+
+                # we failed to parse this string
+                if not requirement_name:
+                    continue
+
+                distribution = self.names_to_distributions.get(requirement_name)
+
+                # it is maybe okay, if we fail to find requirement package on local machine:
+                # 1) we ignoring markers in requirement_string, for example
+                #    "package<1.0.0; python_version<'3.10'"
+                #    so this requirement may be not real for us right here.
+                # 2) we could fail to find it by another reason, for example,
+                #    we are ignoring packages, builded with
+                #    obsolete egg-info; I think it is too much stars to align.
+                if not distribution:
+                    continue
+
+                # if requirement is another meta package, we need to add it to our DFS
+                if check_distribution_is_meta_package(distribution):
+                    if distribution.name not in seen_meta_packages:
+                        meta_packages.add(distribution)
+
+                    continue
+
+                # it is non-meta package, so classify it as usual
+                package = self._classify_distribution(distribution, binary_distributions)
+                result.add(package)
```

Comment:
so method `_classify_meta_package_distributions()` returns not only meta packages, but also usual packages which cannot be explored since they are meta packages dependency?

Reply:
yep
probably I should rename _classify_meta_package_distributions -> _process_meta_packages